feat: script based ae submitter installation for macOS to user preferences directory

jaikish opened the pull request:
Fixes: Internally tracked

### What was the problem/requirement? (What/Why)

Currently `DeadlineCloudAfterEffectsSubmitter` only supports user installation(not Admin). 

In User installation, MacOS AE submitter installer does not install it as an AE script. Customers have to perform manual steps to drag-and-drop from submitter-installed-location to the AE user-preferences(or root) ScriptUI Panels directory. We want to simplify this by making installer do the user-preference installation. 

### What was the solution? (How)

- Add shell scripts for macOS After Effects submitter installation and uninstallation to user preference directories
- Update installer XML configuration to use shell scripts on macOS only
- Focus on user-scope installation approach for macOS

This change needs another change in the shared installer to show the new options and info screens.

### What is the impact of this change?

Customers get the script available in all the current minor versions of AE inside the selected major version.

For example, If the user chooses major version `25` during installation, the `DeadlineCloudSubmitter(User).jsx` and `DeadlineCloudSubmitter_Assets` will be placed inside

```
/Users/<user>/Library/Preferences/Adobe/After Effects/25.0/Scripts/ScriptUI Panels
/Users/<user>/Library/Preferences/Adobe/After Effects/25.2/Scripts/ScriptUI Panels 
/Users/<user>/Library/Preferences/Adobe/After Effects/25.6/Scripts/ScriptUI Panels
```

### How was this change tested?

Tested on mac(new feature) and windows(no regression). This change is backward compatible even without changes in the shared installer package to set the variables.

| Test Matrix                                    | MacOS    | Windows  |
|------------------------------------------------|----------|----------|
| User preference succeeds with IgnoreErrors=Yes | Verified | Verified |
| User preference succeeds with IgnoreErrors=No  | Verified | NA       |
| User Preference fails with IgnoreErrors=Yes    | Verified | NA       |
| User Preference fails with IgnoreErrors=No     | Verified | NA       |

#### If you modified source files, did you run the Python script to update the files under the dist folder?
NA

#### If you modified the `/installer`, `/install_builder`, or `/scripts` logic, please run the integration tests and paste the results below

```
  deadline-cloud-for-after-effects git:(user-install-refactor-4) ✗ hatch run test
=========================================================================== test session starts ===========================================================================
platform darwin -- Python 3.11.10, pytest-8.4.2, pluggy-1.6.0
rootdir: /Users/jaikish/Projects/ae-deadline/deadline-cloud-for-after-effects
configfile: pyproject.toml
plugins: xdist-3.8.0, cov-7.0.0
12 workers [7 items]      
.......                                                                                                                                                             [100%]
=========================================================================== slowest 5 durations ===========================================================================
3.64s call     test/unit/test_font_detection.py::test_missing_font
3.64s call     test/unit/test_font_detection.py::test_font_detection[Calibri]
3.20s call     test/unit/test_font_detection.py::test_font_detection[ArialMT]
3.06s call     test/unit/test_font_detection.py::test_font_detection[TimesNewRomanPSMT]
2.81s call     test/unit/test_font_detection.py::test_font_detection[Helvetica]
============================================================================ 7 pass
```

#### If `installer/` was modified or a file was added/removed from `dist/`, then update the installer tests and post the test results below

```
 hatch run test-installer
==================================================================== test session starts ====================================================================
platform darwin -- Python 3.11.10, pytest-8.4.2, pluggy-1.6.0 -- /Users/jaikish/Library/Application Support/hatch/env/virtual/deadline-cloud-for-after-effects/q_4qnEFG/deadline-cloud-for-after-effects/bin/python
cachedir: .pytest_cache
rootdir: /Users/jaikish/Projects/ae-deadline/deadline-cloud-for-after-effects
configfile: pyproject.toml
plugins: xdist-3.8.0, cov-7.0.0
12 workers [13 items]     
scheduling tests via LoadScheduling

test/installer/test_installer.py::test_did_not_build_with_evaluation_mode 
test/installer/test_installer.py::TestSystemInstall::test_install 
test/installer/test_installer.py::TestSystemInstall::test_uninstall 
test/installer/test_installer.py::TestWindows::test_user_permissions 
test/installer/test_installer.py::TestUserInstall::test_uninstall 
[gw4] [  7%] SKIPPED test/installer/test_installer.py::TestWindows::test_user_permissions 
test/installer/test_installer.py::TestUserInstall::test_install 
[gw8] [ 15%] SKIPPED test/installer/test_installer.py::TestSystemInstall::test_install 
[gw1] [ 23%] SKIPPED test/installer/test_installer.py::test_did_not_build_with_evaluation_mode 
test/installer/test_installer.py::TestLinuxAndMacOS::test_system_permissions 
test/installer/test_installer.py::test_default_location 
test/installer/test_installer.py::TestWindows::test_system_permissions 
[gw5] [ 30%] SKIPPED test/installer/test_installer.py::TestWindows::test_system_permissions 
[gw9] [ 38%] SKIPPED test/installer/test_installer.py::TestSystemInstall::test_uninstall 
[gw3] [ 46%] SKIPPED test/installer/test_installer.py::TestLinuxAndMacOS::test_system_permissions 
test/installer/test_installer.py::TestVerifySigning::test_linux_signing 
[gw11] [ 53%] SKIPPED test/installer/test_installer.py::TestVerifySigning::test_linux_signing 
test/installer/test_installer.py::TestLinuxAndMacOS::test_user_permissions 
test/installer/test_installer.py::TestVerifySigning::test_windows_signing 
[gw10] [ 61%] SKIPPED test/installer/test_installer.py::TestVerifySigning::test_windows_signing 
[gw0] [ 69%] PASSED test/installer/test_installer.py::test_default_location 
test/installer/test_installer.py::TestVerifySigning::test_macos_signing 
[gw0] [ 76%] SKIPPED test/installer/test_installer.py::TestVerifySigning::test_macos_signing 
[gw2] [ 84%] PASSED test/installer/test_installer.py::TestLinuxAndMacOS::test_user_permissions 
[gw6] [ 92%] PASSED test/installer/test_installer.py::TestUserInstall::test_install 
[gw7] [100%] PASSED test/installer/test_installer.py::TestUserInstall::test_uninstall 

==================================================================== slowest 5 durations ====================================================================
4.75s setup    test/installer/test_installer.py::TestUserInstall::test_uninstall
4.66s setup    test/installer/test_installer.py::TestUserInstall::test_install
4.17s setup    test/installer/test_installer.py::TestLinuxAndMacOS::test_user_permissions
2.93s call     test/installer/test_installer.py::test_default_location
2.81s call     test/installer/test_installer.py::TestUserInstall::test_uninstall
```

### Was this change documented?

Documentation will be covered as new UI steps in the shared installer.

### Is this a breaking change?
No

---

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

Commit:
feat: script based ae submitter installation for macOS

- Add shell scripts for macOS After Effects submitter installation and uninstallation
- Update installer XML configuration to use shell scripts on macOS only
- Remove Windows PowerShell scripts and Windows-specific installer logic
- Focus on user-scope installation approach for macOS

Signed-off-by: Jaikish Nunna <jaikish@amazon.com>

jaikish committed:
feat: backward compatibility for script user pref installation

Signed-off-by: Jaikish Nunna <7425471+jaikish@users.noreply.github.com>

viknith commented on install_builder/deadline-cloud-for-after-effects.xml:
```diff
@@ -8,158 +8,143 @@
                     </distributionFileList>
+                    <actionList>
+                        <!-- Make shell scripts executable on macOS -->
+                        <changePermissions permissions="0755" files="${installdir}/scripts/ae_installer_helper.sh">
```
Curious, do these permissions need to set when running on the machine running the installer? or are they file-level permissions that can be set by us before pushing it up to Github?

jaikish replied:
These need to be set while running the installer on the workstation.

viknith commented on install_builder/deadline-cloud-for-after-effects.xml:
```diff
@@ -8,158 +8,143 @@
+                        <!-- Make shell scripts executable on macOS -->
+                        <changePermissions permissions="0755" files="${installdir}/scripts/ae_installer_helper.sh">
+                            <ruleList>
+                                <platformTest type="osx"/>
```
Just checking, was this platform test still needed even though you specified `<platforms>osx</platforms>` above?

jaikish replied:
This is actually redundant now. I had it when I had platform-all when I was shipping ps1 scripts as well in the same bundle. I will remove this check.

viknith commented on install_builder/deadline-cloud-for-after-effects.xml:
```diff
@@ -8,158 +8,143 @@
-                                    <value>${installdir}/AE2025</value>
-                                </setInstallerVariable>
+                                <!-- Set macOS After Effects preferences path -->
+                                <setInstallerVariable name="ae_prefs_path" value="/Users/${system_username}/Library/Preferences/Adobe/After Effects" />
```
I see the parameter list above has a osx restriction, so does that trickle down to here such that this variable is only set on OSX? if so, does there need to be rule list further below on Line 87 or is it redundant? Same question applies for the ae25 section below.

jaikish replied:
It is set to this path but unused on Windows right now, but once we support Windows, the same variable will be set to a different path. I can put a condition for this until we support Windows.

viknith commented on install_builder/deadline-cloud-for-after-effects.xml:
```diff
@@ -179,14 +164,79 @@
+            <name>setup_documentation_explanation</name>
+            <value>This installer installs the Deadline Cloud After Effects submitter in two locations: the "${installdir}/Submitters/AfterEffects/" directory and AE user preferences as "DeadlineCloudSubmitter(User).jsx".
+
+- On the next screen, you can choose to ignore script installation failures in the AE user preference location.
```
I think we spoke about this offline but just for visibility, can you explain further why we're going with this option to ignore script installation failures? When would an artist actual want to specifically ignore script installation failures?

jaikish replied:
The assumption for this PR is that the user preference path exists for all macOS users and they have access to that path. We can't be sure how things work on various enterprise setups. (Maybe network paths, limited access even to home directory, breaking change in how AE handles user prefs etc). 

In any unforeseen runtime failures where our installer is not able to install in user paths, we want to give them the existing functionality of being able to choose a location for the submitter scripts.

viknith commented on scripts/ae_installer_helper.sh:
```diff
@@ -0,0 +1,279 @@
+    files_copied=0
+    copy_failed=false
+    
+    for item in "$SUBMITTER_SRC_PATH"/*; do
```
Nit: having a bit of trouble following the rest of the code under this for-block, admittedly not very experienced reading bash so some more comments explaining what's going on here and in other parts of the file for what checks are being done and why they're happening would be great.

jaikish replied:
Sure! I can do that in next rev

viknith commented:
Some more questions I had:
- Any updates needed to README.md or DEVELOPMENT.md? 
- Were your tests ran on both OSes?
- When you say `Remove Windows PowerShell scripts and Windows-specific installer logic`, what will the Windows install experience look like now? What difference will artists see, do they need to change their behavior during installation to preserve the same experience they had before and is it clear to them on the steps to take to do that?

jaikish commented:
> Some more questions I had:
> 
>     * Any updates needed to README.md or DEVELOPMENT.md?
> 
>     * Were your tests ran on both OSes?
> 
>     * When you say `Remove Windows PowerShell scripts and Windows-specific installer logic`, what will the Windows install experience look like now? What difference will artists see, do they need to change their behavior during installation to preserve the same experience they had before and is it clear to them on the steps to take to do that?


> Any updates needed to README.md or DEVELOPMENT.md?

I added info screens to avoid artists having to go through README. Saying that, we are still retaining the existing experience of placing the script files in user-chosen location or *~/DeadlineCloudSubmitter/AfterEffects/AE*`. That documentation is still relevant.

> Were your tests ran on both OSes?

Manual tests on both OSes, unit and installer tests only on MacOS.


> When you say `Remove Windows PowerShell scripts and Windows-specific installer logic`, what will the Windows install experience look like now? What difference will artists see, do they need to change their behavior during installation to preserve the same experience they had before and is it clear to them on the steps to take to do that?

Windows experience is unchanged. I think the "Remove Windows...." stuff came as I reverted the windows changes in a separate commit due to limitations of powershell integrations in Installbuilder. I will update the commit description.

xuetinp commented on scripts/ae_installer_helper.sh:
```diff
@@ -0,0 +1,279 @@
+    echo "Install Deadline Cloud submitter to After Effects"
+    echo ""
+    echo "Options:"
+    echo "  --installation_scope SCOPE        Installation scope (must be 'user' to perform operations)"
```
Do we need this option if it only accepts one value?

jaikish replied:
Good question!

On MacOS, we only support `user` scope. But on Windows, this script should be executing in `user` scope but skipping in `Admin` mode. (We can put that condition in xml too but I wanted as less branching as possible in the XML).

jaikish replied:
I think I will remove the additional parameter for Mac. Thanks!

jaikish committed:
feat: implemented PR comments

Signed-off-by: Jaikish Nunna <7425471+jaikish@users.noreply.github.com>

viknith commented on scripts/ae_installer_helper.sh:
```diff
@@ -54,20 +83,27 @@ while [[ $# -gt 0 ]]; do
     local error_msg="$1"
     if [[ "$IGNORE_MINOR_VERSION_ERRORS" == "1" ]]; then
+        # GRACEFUL DEGRADATION: Log warning but continue execution
         echo "Warning: $error_msg (ignored due to ignore_minor_version_errors flag)" >&2
```
Curious, can these echoes be found during/after installation?

jaikish replied:
No logging is available as part of the installer but this script can be invoked manually for testing/troubleshooting.

viknith reviewed (approved):
LGTM, great work!

jaikish committed:
Merge branch 'mainline' into user-install-refactor-4
